ceno_zkvm/src/scheme/prover.rs

```diff
@@ -565,49 +565,58 @@ impl<E: ExtensionField, PCS: PolynomialCommitmentScheme<E>> ZKVMProver<E, PCS> {
             }
         };
 
-        let mut virtual_polys = VirtualPolynomials::<E>::new(num_threads, log2_num_instances);
-
         let eq_r = build_eq_x_r_vec(&rt_r[..log2_r_count]);
         let eq_w = build_eq_x_r_vec(&rt_w[..log2_w_count]);
         let eq_lk = build_eq_x_r_vec(&rt_lk[..log2_lk_count]);
 
+        let linear_combine_mles =
+            |coeffs: &[E], mles: &[ArcMultilinearExtension<E>], padding: E| {
+                assert!(!mles.is_empty());
+                assert_eq!(coeffs.len(), mles.len());
+
+                let n = mles[0].evaluations().len();
+                let mle_evals = mles.iter().map(|mle| mle.get_ext_field_vec()).collect_vec();
+                // combine into single mle by dot product with coeff
+                (0..n)
+                    .into_par_iter()
+                    .map(|j| {
+                        dot_product::<E, _, _>(
+                            mle_evals.iter().map(|mle_eval| mle_eval[j] - padding),
+                            coeffs.iter().copied(),
+                        )
+                    })
+                    .collect::<Vec<_>>()
+                    .into_mle()
+                    .into()
+            };
+        let r_records_combined: ArcMultilinearExtension<E> =
+            linear_combine_mles(&eq_r[0..r_counts_per_instance], r_records_wit, E::ONE);
+
+        let w_records_combined: ArcMultilinearExtension<E> =
+            linear_combine_mles(&eq_w[0..w_counts_per_instance], w_records_wit, E::ONE);
+
+        let lk_records_combined: ArcMultilinearExtension<E> = linear_combine_mles(
+            &eq_lk[0..lk_counts_per_instance],
+            lk_records_wit,
+            chip_record_alpha,
+        );
+
+        let mut virtual_polys = VirtualPolynomials::<E>::new(num_threads, log2_num_instances);
+
         // read
         // rt_r := rt || rs
-        for i in 0..r_counts_per_instance {
-            // \sum_t (sel(rt, t) * (\sum_i alpha_read * eq(rs, i) * record_r[t] ))
-            virtual_polys.add_mle_list(vec![&sel_r, &r_records_wit[i]], eq_r[i] * *alpha_read);
-        }
-        // \sum_t alpha_read * sel(rt, t) * (\sum_i (eq(rs, i)) - 1)
-        virtual_polys.add_mle_list(
-            vec![&sel_r],
-            *alpha_read * eq_r[r_counts_per_instance..].iter().copied().sum::<E>() - *alpha_read,
-        );
+        // \sum_t alpha_read * sel(rt, t) * (\sum_{i < r_counts} eq(rs, i) * (record_r[t] - 1))
```

Comment:
nitpick: probably we can move comment to the linear mle combination part

Reply:
I've add a few comments to explain the selector sumcheck. https://github.com/scroll-tech/ceno/pull/931/commits/14e98bdb018e261c12f04ba8e10f8c3b0bd4ad96#diff-c7a9e29146c36dca3fe9a0e5dd50eabbaad0e8aba21a2517fe6105eb9212058eR594